src/Microsoft.ML.Tokenizers/Model/SentencePieceBpe.cs

```diff
@@ -0,0 +1,952 @@
+        private Dictionary<string, int>? _publicVocab;
+        private readonly int _maxByteId;
+        private readonly int _byteCodeToIdOffset; // offset of mapping byte code to the to the Ids.
+        private readonly int _oneByteUtf8EncodingMaxId; // the maximum value of the one byte UTF-8 character.
+
+        internal SentencePieceBpe(ModelProto modelProto, bool addBos, bool addEos) :
+                this(modelProto is null ? throw new ArgumentNullException(nameof(modelProto)) : modelProto)
+        {
+            AddBeginningOfSentence = addBos;
+            AddEndOfSentence = addEos;
+        }
+
+        private SentencePieceBpe(ModelProto modelProto)
+        {
+            for (int i = 0; i < modelProto.Pieces.Count; i++)
+            {
+                var piece = modelProto.Pieces[i];
+                _vocab.Add(new StringSpanOrdinalKey(piece.Piece), (i, piece.Score, (byte)piece.Type));
+                _vocabReverse.Add(i, piece.Piece);
+
+                if (piece.Type == ModelProto.Types.SentencePiece.Types.Type.Byte)
+                {
+                    _maxByteId = i;
+                }
+            }
+
+            _byteCodeToIdOffset = _vocab.TryGetValue("<0x00>", out (int Id, float Score, byte Type) value) ? value.Id : _maxByteId;
+            _oneByteUtf8EncodingMaxId = _byteCodeToIdOffset + 0x7F; // 0x7F is the maximum value of the one byte UTF-8 character.
+
+            BeginningOfSentenceToken = modelProto.TrainerSpec.BosPiece ?? "<s>";
+            BeginningOfSentenceId = modelProto.TrainerSpec.BosId <= 0 ? 1 : modelProto.TrainerSpec.BosId;
+            EndOfSentenceToken = modelProto.TrainerSpec.EosPiece ?? "</s>";
+            EndOfSentenceId = modelProto.TrainerSpec.EosId <= 0 ? 1 : modelProto.TrainerSpec.EosId;
+            UnknownToken = modelProto.TrainerSpec.UnkPiece ?? "<unk>";
+            UnknownId = modelProto.TrainerSpec.UnkId < 0 ? 0 : modelProto.TrainerSpec.UnkId;
+
+            AddDummyPrefix = modelProto.NormalizerSpec.AddDummyPrefix;
+            EscapeWhiteSpaces = modelProto.NormalizerSpec.EscapeWhitespaces;
+            TreatWhitespaceAsSuffix = modelProto.TrainerSpec.TreatWhitespaceAsSuffix;
+            ByteFallback = modelProto.TrainerSpec.ByteFallback;
+        }
+
+        /// <summary>
+        /// Create a SentencePieceBpe tokenizer from the given model stream. The model stream should contain the SentencePiece Bpe model according to
+        /// https://github.com/google/sentencepiece/blob/master/src/sentencepiece_model.proto specification.
+        /// </summary>
+        /// <param name="modelStream">The stream containing the SentencePiece Bpe model.</param>
+        /// <param name="addBeginOfSentence">Indicate emitting the beginning of sentence token during the encoding.</param>
+        /// <param name="addEndOfSentence">Indicate emitting the end of sentence token during the encoding.</param>
+        public static Tokenizer CreateLlamaTokenizer(
```

Comment:
I'm not sure this is the best place to have this. Maybe we should have a separate class called Llama? Someone trying to find this would need to know that sentence piece is what's used for Llama.

Reply:
Or, maybe this should just be called CreateTokenizer?

Reply:
I didn't call it CreateTokenizer because it has some limitation we don't support in general with this model, like if the data file support normalization and we'll throw on that. with Llama, we know not using normalization. The files I used in my testing is actual data files used with the Llama tokenizer. Also, I didn't expose a whole class just to expose the tokenizer from. What about move this method to the Tokenizer class itself, discoverability will be better I guess there. what do you think?